modules/distributor/distributor.go

```diff
@@ -530,19 +561,54 @@ func requestsByTraceID(batches []*v1.ResourceSpans, userID string, spanCount int
 	return keys, traces, nil
 }
 
-func recordDiscaredSpans(err error, userID string, spanCount int) {
-	s := status.Convert(err)
-	if s == nil {
-		return
+func countDiscaredSpans(numSuccessByTraceIndex []int, lastErrorReasonByTraceIndex []tempopb.PushErrorReason, traces []*rebatchedTrace, repFactor int) (maxLiveDiscardedCount int, traceTooLargeDiscardedCount int) {
+	quorum := int(math.Floor(float64(repFactor)/2)) + 1 // min success required
+
+	for traceIndex, numSuccess := range numSuccessByTraceIndex {
+		// we will count anything that did not receive min success as discarded
+		if numSuccess < quorum {
+			spanCount := traces[traceIndex].spanCount
+			switch lastErrorReasonByTraceIndex[traceIndex] {
+			case tempopb.PushErrorReason_MAX_LIVE_TRACES:
+				maxLiveDiscardedCount += spanCount
+			case tempopb.PushErrorReason_TRACE_TOO_LARGE:
+				traceTooLargeDiscardedCount += spanCount
+			}
+		}
 	}
-	desc := s.Message()
 
-	if strings.HasPrefix(desc, overrides.ErrorPrefixLiveTracesExceeded) {
-		overrides.RecordDiscardedSpans(spanCount, reasonLiveTracesExceeded, userID)
-	} else if strings.HasPrefix(desc, overrides.ErrorPrefixTraceTooLarge) {
-		overrides.RecordDiscardedSpans(spanCount, reasonTraceTooLarge, userID)
+	return maxLiveDiscardedCount, traceTooLargeDiscardedCount
+}
+
+func (d *Distributor) processPushResponse(pushResponse *tempopb.PushResponse, numSuccessByTraceIndex []int, lastErrorReasonByTraceIndex []tempopb.PushErrorReason, numOfTraces int, indexes []int) {
```

Comment:
this function needs some cleanup and tests. going to make some stylistic suggestions.